### PR TITLE
feat(business_profile): introduce domain models for business profile v1 and v2 APIs

### PR DESCRIPTION
**Pull request by SanchithHegde**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR includes the following changes:

1. Refactors code to ensure that all updates to the `business_profile` table always include an update of the `modified_at` field.
2. Introduces the `BusinessProfileUpdate::RoutingAlgorithmUpdate` for use in routing code and to slightly improve code readability.
3. Updates the `apply_changeset()` method on `BusinessProfileUpdate` to use fields from the source object for fields which are not being updated. 
   - This function is only called in the `BusinessProfileInterface` implementation for `MockDb`, so **this should not affect production code**.
4. Adds database migrations for business profile v2, and introduces corresponding types (and feature flags) in the `diesel_models` and `common_enums` crates.
5. Introduces business profile v1 and v2 types (and feature flags) in the `hyperswitch_domain_models` crate.
   - The code has not been updated to make use of either of these domain models anywhere. The refactor to use business profile v1 domain models throughout the codebase will be taken up in a separate PR, since it affects a lot of files.

To reiterate, **this PR does not affect any production code**, and only adds new code, while moving existing code behind suitable feature flags.

### Additional Changes

- [ ] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

This PR includes database migrations for and updates schema for business profile v2 APIs, does not affect business profile v1 schema.

## Motivation and Context

Enabling / unblocking business profile v2 changes or any other changes dependent on business profile v2 types.

## How did you test it?

Ran the following commands and ensured they've passed:

```shell
# Ensuring successful compilation of v1 and v2 diesel models
cargo clippy -p diesel_models
cargo clippy -p diesel_models --no-default-features --features v2,business_profile_v2

# Ensuring successful compilation of v1 and v2 domain models
cargo clippy -p hyperswitch_domain_models
cargo clippy -p hyperswitch_domain_models --no-default-features --features v2,business_profile_v2

# Ensuring clippy passes as usual
just clippy
```

Attaching a screenshot of the same:
Screenshot of checks running successfully

In addition, I've done sanity testing using Postman.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: justfile [hunk 1]
```diff
@@ -120,7 +120,7 @@ euclid-wasm features='dummy_connector':
 precommit: fmt clippy
 
 # Check compilation of v2 feature on base dependencies
-v2_intermediate_features := "merchant_account_v2,payment_v2,customer_v2"
+v2_intermediate_features := "merchant_account_v2,payment_v2,customer_v2,business_profile_v2"
 hack_v2:
     cargo hack clippy --feature-powerset --ignore-unknown-features --at-least-one-of "v2 " --include-features "v2" --include-features {{ v2_intermediate_features }} --package "hyperswitch_domain_models" --package "diesel_models" --package "api_models"
```
Comment (Narayanbhat166) [inline justfile:125]: can we not have the feature-set check, we can check each create with independent v2 features alone. Can be taken up in a separate PR which includes only this change

**Review by Narayanbhat166 — APPROVED** [review]

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+    pub profile_id: String,
+    pub merchant_id: common_utils::id_type::MerchantId,
+    pub profile_name: String,
+    pub created_at: time::PrimitiveDateTime,
+    pub modified_at: time::PrimitiveDateTime,
+    pub return_url: Option<String>,
+    pub enable_payment_response_hash: bool,
+    pub payment_response_hash_key: Option<String>,
+    pub redirect_to_merchant_with_http_post: bool,
+    pub webhook_details: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:399]:
```suggestion
    pub webhook_details: Option<pii::SecretSerdeValue>,
```

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+    pub metadata: Option<pii::SecretSerdeValue>,
+    pub is_recon_enabled: bool,
+    pub applepay_verified_domains: Option<Vec<String>>,
+    pub payment_link_config: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:403]:
```suggestion
    pub payment_link_config: Option<pii::SecretSerdeValue>,
```

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+    pub session_expiry: Option<i64>,
+    pub authentication_connector_details: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:405]:
```suggestion
    pub authentication_connector_details: Option<pii::SecretSerdeValue>,
```

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 5]
```diff
@@ -0,0 +1,763 @@
...
+    pub payout_link_config: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:406]:
```suggestion
    pub payout_link_config: Option<pii::SecretSerdeValue>,
```

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+    pub use_billing_as_payment_method_billing: Option<bool>,
+    pub collect_shipping_details_from_wallet_connector: Option<bool>,
+    pub collect_billing_details_from_wallet_connector: Option<bool>,
+    pub outgoing_webhook_custom_http_headers: OptionalEncryptableValue,
+    pub routing_algorithm_id: Option<String>,
+    pub order_fulfillment_time: Option<i64>,
+    pub order_fulfillment_time_origin: Option<OrderFulfillmentTimeOrigin>,
+    pub frm_routing_algorithm_id: Option<String>,
+    pub payout_routing_algorithm_id: Option<String>,
+    pub default_fallback_routing: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:419]:
```suggestion
    pub default_fallback_routing: Option<pii::SecretSerdeValue>,
```

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]
